doc: engine version enforcement (#766)
longhorn/website: content/docs/1.6.0/deploy/important-notes/index.md

The CSI components in Longhorn v{{< current-version >}} only function with the `v1` version.
Please follow the instructions at [Enable CSI Snapshot Support](../../snapshots-and-backups/csi-snapshot-support/enable-csi-snapshot-support) to update CSI snapshot CRDs and the CSI snapshot controller.
If you have Longhorn volume manifests or scripts that are still using `v1beta1` version, you must upgrade them to `v1` as well.

### Engine Upgrade Enforcement

Beginning with version v1.6.0, Longhorn is implementing mandatory engine upgrades. See the [release note](https://github.com/longhorn/longhorn/releases/tag/v{{< current-version >}}) for information about the minimum supported engine image version.

When upgrading through Helm, a component compatibility check is automatically performed. If the new Longhorn is not compatible with the engine images that are currently in use, the upgrade path is blocked through a pre-hook mechanism.

> If you installed Longhorn using the manifests, engine upgrades are enforced by Longhorn Manager. Attempts to upgrade Longhorn Manager may cause unsuccessful pod launches and generate corresponding error logs. If you encounter such errors, you must revert to the previous Longhorn version and then upgrade the engines that are using the incompatible engine images.

Reviewer: ```suggestion
If you installed Longhorn using the manifests, engine upgrades are enforced by the Longhorn Manager. Attempts to upgrade Longhorn Manager may cause unsuccessful pod launches and generate corresponding error logs, although it poses no harm. If you encounter such errors, you must revert to the previous Longhorn version and then upgrade the engines that are using the incompatible engine images before the next upgrade.
```

The CSI components in Longhorn v{{< current-version >}} only function with the `v1` version.
Please follow the instructions at [Enable CSI Snapshot Support](../../snapshots-and-backups/csi-snapshot-support/enable-csi-snapshot-support) to update CSI snapshot CRDs and the CSI snapshot controller.
If you have Longhorn volume manifests or scripts that are still using `v1beta1` version, you must upgrade them to `v1` as well.

### Engine Upgrade Enforcement

Beginning with version v1.6.0, Longhorn is implementing mandatory engine upgrades. See the [release note](https://github.com/longhorn/longhorn/releases/tag/v{{< current-version >}}) for information about the minimum supported engine image version.

When upgrading through Helm, a component compatibility check is automatically performed. If the new Longhorn is not compatible with the engine images that are currently in use, the upgrade path is blocked through a pre-hook mechanism.

> If you installed Longhorn using the manifests, engine upgrades are enforced by the Longhorn Manager. Attempts to upgrade Longhorn Manager may cause unsuccessful pod launches and generate corresponding error logs, although it poses no harm. If you encounter such errors, you must revert to the previous Longhorn version and then upgrade the engines that are using the incompatible engine images before the next upgrade.

Reviewer 2: ```suggestion
If you installed Longhorn using a manifest file, engine upgrades are enforced by Longhorn Manager. Attempts to upgrade Longhorn Manager may cause unsuccessful pod launches (that pose no harm to the system) and generate corresponding error logs. If you encounter such errors, you must revert to the previous Longhorn Manager version and then upgrade engines with incompatible engine images.
```

Author: I've updated the top section of the suggestion. However, the lower part looks a bit odd to me.

> you must revert to the previous Longhorn Manager version

The user should revert the Longhorn version, not just the manager's version.

> upgrade engines with incompatible engine images

This sounds like we are asking the user to upgrade the engine using an incompatible engine image.

> If you installed Longhorn using a manifest file, engine upgrades are enforced by Longhorn Manager. Attempts to upgrade Longhorn Manager may cause unsuccessful pod launches (that pose no harm to the system) and generate corresponding error logs. If you encounter such errors, you must revert to the previous Longhorn version and then upgrade the engines that are using the incompatible engine images before the next upgrade.

Reviewer 2: ```suggestion
If you installed Longhorn using a manifest file, engine upgrades are enforced by Longhorn Manager. Attempts to upgrade Longhorn Manager may cause unsuccessful pod launches (that pose no harm to the system) and generate corresponding error logs. If you encounter such errors, you must revert to the previous Longhorn version and then upgrade the engines that are using the incompatible engine images before the next upgrade.
```

Author: The suggestion looks the same. I've reverted the last change.

- BackingImageManager.status.backingImageFileMap.directory
- BackingImageManager.status.backingImageFileMap.downloadProgress
- BackingImageManager.status.backingImageFileMap.url

### Engine Upgrade Enforcement

> Starting from `v1.6.0`, Longhorn will implement mandatory version upgrades for the engines. The minimum required supporting version will be outlined in the [release note](https://github.com/longhorn/longhorn/releases/tag/v{{< current-version >}}).

Reviewer 2: Beginning with version `v1.6.0`, Longhorn will implement mandatory engine upgrades. See the [release note](https://github.com/longhorn/longhorn/releases/tag/v{{< current-version >}}) for information about the minimum supported engine image version.

> Starting from `v1.6.0`, Longhorn will implement mandatory version upgrades for the engines. The minimum required supporting version will be outlined in the [release note](https://github.com/longhorn/longhorn/releases/tag/v{{< current-version >}}).

Reviewer 2: ```suggestion
Beginning with version v1.6.0, Longhorn is implementing mandatory engine upgrades. See the [release note]([https://github.com/longhorn/longhorn/releases/tag/v{{](https://github.com/longhorn/longhorn/releases/tag/v%7B%7B)< current-version >}}) for information about the minimum supported engine image version.
```

Author: ![Screenshot from 2023-10-18 16-59-04](https://github.com/longhorn/website/assets/22583541/0cae3e38-6748-412c-b871-2ac6931d7c4b)

The format is a bit off, but I've updated your suggested words.

Reviewer 2: I made a mistake in copying/pasting the URL that resulted in extra characters. Sorry about that.

> When performing a helm upgrade, incompatible versions are automatically detected and prevented through a pre-hook mechanism. However, if you've installed Longhorn using manifests, the enforcement is handled by the longhorn-manager. This will result in unsuccessful pod launches and corresponding error logs. If you encounter such an error, it's necessary to revert to the previous version and upgrade any engine currently utilizing the incompatible engine image.

Reviewer 2: When upgrading through Helm, a component compatibility check is automatically performed. If the new Longhorn Manager version is not compatible with the engine images that are currently in use, the upgrade path is blocked through a pre-hook mechanism.

If you installed Longhorn using a manifest file, engine upgrades are enforced by Longhorn Manager. Attempts to upgrade Longhorn Manager may cause unsuccessful pod launches and generate corresponding error logs. If you encounter such errors, you must revert to the previous Longhorn Manager version and then upgrade engines using incompatible engine images.

> When performing a helm upgrade, incompatible versions are automatically detected and prevented through a pre-hook mechanism. However, if you've installed Longhorn using manifests, the enforcement is handled by the longhorn-manager. This will result in unsuccessful pod launches and corresponding error logs. If you encounter such an error, it's necessary to revert to the previous version and upgrade any engine currently utilizing the incompatible engine image.

Reviewer 2: ```suggestion
When upgrading through Helm, a component compatibility check is automatically performed. If the new Longhorn Manager is not compatible with the engine images that are currently in use, the upgrade path is blocked through a pre-hook mechanism.

If you installed Longhorn using a manifest file, engine upgrades are enforced by Longhorn Manager. Attempts to upgrade Longhorn Manager may cause unsuccessful pod launches and generate corresponding error logs. If you encounter such errors, you must revert to the previous Longhorn Manager version and then upgrade engines using incompatible engine images.
```

Author: I like the suggestions, but there are some parts that seem a little confusing. For instance:
- `If the new Longhorn Manager is not compatible with the engine images that are currently in use`. We've implemented the check in the Longhorn manager to ensure compatibility across the entire Longhorn system.
- `you must revert to the previous Longhorn Manager version and then upgrade engines using incompatible engine images.` This is a bit confusing because we don't actually upgrade the engines using the engine images.

So based on your suggestions, I've modified it a little. PTAL.

Reviewer 2: Thanks for clarifying. Your changes look good.

> You can assess the engine image versions for your current engines by utilizing the script below:

Reviewer 2: You can determine the versions of engine images that are currently in use with the following script:

```bash
#!/bin/bash

namespace="longhorn-system"

engine_images=$(kubectl -n $namespace get engineimage -o=jsonpath='{.items[*].metadata.name}')

for engine_image in $engine_images; do
    cli_api_version=$(kubectl -n $namespace get engineimage $engine_image -o=jsonpath='{.status.cliAPIVersion}')
    controller_api_version=$(kubectl -n $namespace get engineimage $engine_image -o=jsonpath='{.status.controllerAPIVersion}')
    echo "EngineImage: $engine_image | cliAPIVersion: $cli_api_version | controllerAPIVersion: $controller_api_version"
done
```

> After successfully upgrading to version 1.6.0, the engine image versions will be accessible in the UI for subsequent version upgrades.

Reviewer 2: Suggested wording: 
Once you successfully upgrade to version `v1.6.0`, you will be able to view information about engine image versions on the UI.
